src/Servant/Reflex/Response.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE ConstraintKinds   #-}
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Response handlers
+--   Usefull for 'clientWithOptsAndResultHandler'
+module Servant.Reflex.Response
+  ( reloadOnAPIError
+  , reloadOnResponse
+  , RefreshSupport
+  , isAuthErr
+  , is4xxErr
+  , isDecodeError
+  ) where
+
+import           Control.Monad               (join, void)
+import qualified Data.Text                   as Text
+import           JSDOM                       (currentDocument)
+import           JSDOM.Generated.Document    (getLocation)
+import           JSDOM.Types                 (Location, liftDOM)
+import           Language.Javascript.JSaddle (MonadJSM, ( # ))
+import           Reflex
+import           Reflex.Dom.Builder.Class
+import           Reflex.Dom.Widget.Basic
+import           Reflex.Dom.Xhr
+import           Servant.Reflex
+
+-- | Alias for constraints that can refresh
+type RefreshSupport t m
+   = (MonadJSM m, MonadHold t m, DomBuilder t m, SupportsServantReflex t m)
+
+is4xxErr :: ReqResult () a -> Bool
+is4xxErr =
+  maybe
+    False
+    (\req ->
+       (_xhrResponse_status req < fromInteger 500) &&
+       (_xhrResponse_status req >= 400)) .
+  response
+
+isAuthErr :: ReqResult () a -> Bool
+isAuthErr =
+  maybe
+    False
+    (\req ->
+       let stat = _xhrResponse_status req
+       in (stat == fromInteger 401) &&
+           (stat == fromInteger 402) && (stat == fromInteger 407)) .
+  response
+
+isDecodeError :: ReqResult () a -> Bool -- TODO find a better way of detecting this, is there no typesafe decode error?
+isDecodeError = maybe False (Text.isInfixOf "Error in $") . reqFailure
+
+-- | Reloads page and clears the cache on any 4xx response which is not auth related
+--   or on any decoding errors.
+--   Most endpoints want to have this.
+--   These errors mean the client code is probably outdated.
+--   This assumes client errros aren't used in your app logic,
+--   for more flexibility see 'reloadOnResponse'
+reloadOnAPIError ::
+     (MonadJSM m, RefreshSupport t m)
+  => Event t (ReqResult () a)
+  -> m (Event t (ReqResult () a))
+reloadOnAPIError =
+  reloadOnResponse
+    (\req -> (is4xxErr req && (not $ isAuthErr req)) || isDecodeError req)
+
+-- | Reload and clears the cache if the predicate returns true
+reloadOnResponse ::
+     (MonadJSM m, RefreshSupport t m)
+  => (ReqResult () a -> Bool)
+  -> Event t (ReqResult () a)
+  -> m (Event t (ReqResult () a))
+reloadOnResponse predic reqResult = do
+  void $
+    widgetHold blank $
+    ffor reqResult $ \req ->
+      if predic req
+        then clearCacheReload
+        else pure ()
+  pure reqResult
+
+-- | <https://developer.mozilla.org/en-US/docs/Web/API/Location.reload Mozilla Location.reload documentation>
+reload' :: MonadJSM m => Location -> Bool -> m ()
+reload' self bool = liftDOM $ void $ self # funName $ bool
```

Comment:
Rather than just reloading, it may be a nicer dev experience if there were a modal popping up, saying "There was an unexpected 4xx error. Reload?", and offering a Reload button (one that clears the cache, as you do here). That could help users break out of accidental infinite-reload cycles.

Reply:
I could see this work with javascript alerts: http://hackage.haskell.org/package/jsaddle-dom-0.9.3.1/docs/JSDOM-Generated-Window.html#v:alert

For my own codebase I just assume I wouldn't get into a loop. I've found that having stale client code happens so often that it's worth the tradeoff.